sei-tendermint/internal/p2p/pex/reactor.go

```diff
@@ -83,345 +77,141 @@ func ChannelDescriptor() conn.ChannelDescriptor {
-
-	// the total number of unique peers added
-	totalPeers int
-
-	channel *p2p.Channel
-
-	// Used to signal a restart the node on the application level
-	restartCh                     chan<- struct{}
-	restartNoAvailablePeersWindow time.Duration
+	sendInterval time.Duration
+	logger       log.Logger
+	router       *p2p.Router
+	// peerLimiters limits the number of messages received from peers.
+	peerLimiters utils.Mutex[map[types.NodeID]*rate.Limiter]
+	channel      *p2p.Channel
 }
 
 // NewReactor returns a reference to a new reactor.
 func NewReactor(
 	logger log.Logger,
 	router *p2p.Router,
-	restartCh chan<- struct{},
-	selfRemediationConfig *config.SelfRemediationConfig,
+	sendInterval time.Duration,
 ) (*Reactor, error) {
 	channel, err := router.OpenChannel(ChannelDescriptor())
 	if err != nil {
 		return nil, err
 	}
 	r := &Reactor{
-		logger:                        logger,
-		channel:                       channel,
-		router:                        router,
-		availablePeers:                make(map[types.NodeID]struct{}),
-		lastNoAvailablePeers:          time.Time{},
-		requestsSent:                  make(map[types.NodeID]struct{}),
-		lastReceivedRequests:          make(map[types.NodeID]time.Time),
-		restartCh:                     restartCh,
-		restartNoAvailablePeersWindow: time.Duration(selfRemediationConfig.P2pNoPeersRestarWindowSeconds) * time.Second,
+		logger:       logger,
+		sendInterval: sendInterval,
+		channel:      channel,
+		router:       router,
+		peerLimiters: utils.NewMutex(map[types.NodeID]*rate.Limiter{}),
 	}
-
 	r.BaseService = *service.NewBaseService(logger, "PEX", r)
 	return r, nil
 }
 
-// OnStart starts separate go routines for each p2p Channel and listens for
-// envelopes on each. In addition, it also listens for peer updates and handles
-// messages on that p2p channel accordingly. The caller must be sure to execute
-// OnStop to ensure the outbound p2p Channels are closed.
+func (r *Reactor) Run(ctx context.Context) error {
+	return scope.Run(ctx, func(ctx context.Context, s scope.Scope) error {
+		s.SpawnNamed("sendRoutine", func() error { return r.sendRoutine(ctx) })
+		s.SpawnNamed("processPexCh", func() error { return r.processPexCh(ctx) })
+		s.SpawnNamed("processPeerUpdates", func() error { return r.processPeerUpdates(ctx) })
+		return nil
+	})
+}
+
+// OnStart implements service.Service.
 func (r *Reactor) OnStart(ctx context.Context) error {
-	r.Spawn("processPexCh", func(ctx context.Context) error { return r.processPexCh(ctx) })
-	r.Spawn("processPeerUpdates", func(ctx context.Context) error { return r.processPeerUpdates(ctx) })
+	r.SpawnCritical("Run", func(ctx context.Context) error { return r.Run(ctx) })
 	return nil
 }
 
-// OnStop stops the reactor by signaling to all spawned goroutines to exit and
-// blocking until they all exit.
+// OnStop implements service.Service.
 func (r *Reactor) OnStop() {}
 
+func (r *Reactor) sendRoutine(ctx context.Context) error {
```

Comment:
Do we have any concerns on whether this will cause the network adapter to be utilized/blocked as we broadcast to a batch of peers (eg. 100). Is it possible here that every 10 seconds, we cause a slowdown to other network communications (eg. consensus voting - maybe this specific message type isnt as problematic bc priority) or should it be fine?